### PR TITLE
Tag completion in Emacs 25

### PR DESCRIPTION
**Pull request by johanclaesson**
Adapt tag completion for changes in Emacs 25.

Stop storing module:tag entries in the completion table for definitions
in hrl-files.  These entries make no sense since hrl-files are not
modules.

Add test-case for tag completion in erlang-test.el.

Ensure prog-mode is defined for the sake of Emacs 23 and XEmacs.

Add target tags to top level Makefile.

### CODE REVIEW

File: Makefile.in [hunk 1]
```diff
@@ -1121,3 +1121,8 @@ bootstrap_clean:
 		|| $(MAKE) BOOTSTRAP_ROOT=$(BOOTSTRAP_ROOT) bootstrap_root_clean
 
 # ----------------------------------------------------------------------
+
+.PHONY: TAGS tags
+
+TAGS tags:
+	find . -type f -name "*.[eh]rl" -o -name "*.[ch]" | etags -o TAGS -
```
Comment (psyeugenic) [inline Makefile.in]: No.

**Comment by dgud** [comment]
@psyeugenic the grumpy vim user :-)

**Comment by dgud** [comment]
This PR is already merged to our internal master, not yet pushed to github.  
Can you make a new PR for this commit.

**Comment by johanclaesson** [comment]
Completion in erlang shell buffer do not work in Emacs 24 because of a bug 
in  tags-completion-at-point-function.  That bug is corrected in Emacs 25.
To make it work in Emacs 24 one could apply the following correction:

(when (< emacs-major-version 25)
  (require 'etags)
  (defun tags-completion-at-point-function ()
    "Using tags, return a completion table for the text around point.
If no tags table is loaded, do nothing and return nil."
    (when (or tags-table-list tags-file-name)
      (let ((completion-ignore-case (if (memq tags-case-fold-search '(t nil))
                                        tags-case-fold-search
                                      case-fold-search))
            (pattern (funcall (or find-tag-default-function
                                  (get major-mode 'find-tag-default-function)
                                  'find-tag-default)))
            beg)
        (when pattern
          (save-excursion
            (forward-char (min (- (point-max) (point))
                               (1- (length pattern))))
            (search-backward pattern)
            (setq beg (point))
            (forward-char (length pattern))
            (list beg (point) (tags-lazy-completion-table)
                  :exclusive 'no)))))))

But since this is redefining a built-in function it should probably not 
be included in erlang.el.

**Comment by johanclaesson** [comment]
Sure i will make a new PR.

**Comment by dgud** [comment]
Thanks, merged.

**Comment by johanclaesson** [comment]
Thanks.  Here is the new PR:
https://github.com/erlang/otp/pull/1106

**Pull request closed without merging**